library/ccm.c

```diff
@@ -422,27 +431,32 @@ int mbedtls_ccm_self_test( int verbose )
         if( verbose != 0 )
             mbedtls_printf( "  CCM-AES #%u: ", (unsigned int) i + 1 );
 
+        memset( plaintext, 0, CCM_SELFTEST_PT_MAX_LEN );
+        memset( cipertext, 0, CCM_SELFTEST_CT_MAX_LEN );
+        memcpy( plaintext, msg, msg_len[i]);
```

Comment:
An alternative approach to copying would be to add alignment directives to the compiler for the static buffers that hold the test data. This would save stack space, since the data would already be aligned properly.

Reply:
I tried using thew __align__ attribute. 
1. This didn't work
2. This is toolchain dependent

Reply:
1. Why not? What's the memory layout look like?
2. I had a quick look through the code base and was surprised to find no macro to align a buffer, defined different ways based on the toolchain. This would be useful to have.

Reply:
@RonEld: I reviewed our internal conversation in IOTSSL-1316 once more. After taking a look at this patch, I still do not think that the problems you are experiencing with the accelerators are due to either of the reasons cited:

1. alignment: Indeed, I expect that many accelerators will want their input aligned to some boundary (e.g word, double word, halfword, etc). However, you also mentioned that the accelerator that failed for you did not succeed when you added the alignment attribute to the arrays. To me, this indicates that the problem is not due to alignment issues. Also, I had a quick look at the objdump of a compilation for one of the embedded targets and found that the arrays are all aligned to word boundaries already (which is what you wanted for your specific case). In fact, I expect the compiler to be aware that most devices work better in aligned data, so they insert padding in structs, etc to improve performance. Actually, i expect its not that straight forward to make it not pad stuff, for example there are attributes that you can set to coerce it into packing (not padding) stuff. Just to experiment I added 2 bytes to the `iv` and 1 byte to the `ad` array and found that the compiler (arm-none-eabi-gcc) aligned both of them to a word boundary by putting some `*fill*` component. Here are the contents of the map file for reference:
```
 .rodata.iv     0x0000000000049224        0xe BUILD/NRF52840_DK/GCC_ARM/mbed-os/features/mbedtls/src/ccm.o
 *fill*         0x0000000000049232        0x2 <---- PADDING ADDED BY COMPILER
 .rodata.iv_len
                0x0000000000049234        0xc BUILD/NRF52840_DK/GCC_ARM/mbed-os/features/mbedtls/src/ccm.o
 .rodata.ad     0x0000000000049240       0x15 BUILD/NRF52840_DK/GCC_ARM/mbed-os/features/mbedtls/src/ccm.o
 *fill*         0x0000000000049255        0x3 <---- PADDING ADDED BY COMPILER
 .rodata.aes_info
                0x0000000000049258       0x1c BUILD/NRF52840_DK/GCC_ARM/mbed-os/features/mbedtls/src/cipher_wrap.o
```

2. non-continuous memory: I do not think this is the problem. If the compiler could break up arrays and make them non-contiguous a lot more things would be failing because there are many places in the code that rely on plain pointer arithmetic, which would not work if the data was not contiguous.

Instead, my theory is that the accelerators are probably failing because they might have limitations on the memory ranges they can operate on. For example, in the past we worked with some hardware whose crypto accelerator could only access memory in the 0x2xxxxxxx. It is possible that something similar is happening on your target and that this situation is not being checked by the driver code. What do you think of this? I tried checking the documentation for your hardware, but unfortunately I could not get a hold of it.

On a separate note, IOTSSL-1316 also mentions that even after moving the arrays with the test data onto the stack, there were some ffailures still. Does this patch solve those problems too?